### PR TITLE
Implement basic enum formatting

### PR DESCRIPTION
**Pull request by marcusklaas**
This PR adds formatting for enums with tuple-like variants.

cc https://github.com/nrc/rustfmt/issues/76

### CODE REVIEW

File: tests/idem/enum.rs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    B = 2, // comment
+    C,
+}
+
+enum LongVariants {
+    First(LOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOONG, // small comment
+          VARIANT),
+    // This is the second variant
+    Second,
+}
```
Comment (nrc) [inline tests/idem/enum.rs]: Could you add an attribute or doc comment to one of the variants please? This is often a tricky edge case.

Comment (marcusklaas) [inline tests/idem/enum.rs]: Sure, they seem to work OK.

File: src/items.rs [hunk 2]
```diff
@@ -408,6 +408,122 @@ impl<'a> FmtVisitor<'a> {
...
+
+            self.visit_variant(f, i == enum_def.variants.len() - 1, next_span_start);
+        }
+        self.block_indent -= config!(tab_spaces);
+
+        self.format_missing_with_indent(span.lo + BytePos(enum_snippet.rfind('}').unwrap() as u32));
+        self.changes.push_str_span(span, "}");
+    }
+
+    // Variant of a enum
```
Comment (nrc) [inline src/items.rs]: nit: "an"

File: src/items.rs [hunk 3]
```diff
@@ -408,6 +408,122 @@ impl<'a> FmtVisitor<'a> {
...
+            return;
+        }
+
+        if let ast::VariantKind::TupleVariantKind(ref types) = field.node.kind {
+            self.format_missing_with_indent(field.span.lo);
+
+            let vis = match field.node.vis {
+                ast::Visibility::Public => "pub ",
+                ast::Visibility::Inherited => ""
+            };
```
Comment (nrc) [inline src/items.rs]: We should factor this out into a utility function

File: src/items.rs [hunk 4]
```diff
@@ -408,6 +408,122 @@ impl<'a> FmtVisitor<'a> {
...
+
+            let mut header = vis.to_owned();
+            header.push_str(&field.node.name.to_string());
```
Comment (nrc) [inline src/items.rs]: Probably better to push 3 times to self.changes, rather than pushing to header and then pushing header (I realise I probably did this in other places, but I think it would be better not to, with hindsight)

File: src/items.rs [hunk 5]
```diff
@@ -408,6 +408,122 @@ impl<'a> FmtVisitor<'a> {
...
+                    trailing_separator: SeparatorTactic::Never,
+                    indent: indent,
+                    h_width: budget,
+                    v_width: budget,
+                };
+                result.push_str(&write_list(&type_strings, &fmt));
+                result.push(')');
+            }
+
+            if let Some(ref expr) = field.node.disr_expr {
```
Comment (nrc) [inline src/items.rs]: Do we need to account for this in the budget for the types above? If not we should assert! something here.

File: src/items.rs [hunk 6]
```diff
@@ -408,6 +408,122 @@ impl<'a> FmtVisitor<'a> {
...
+                result.push_str(" = ");
+                result.push_str(&self.snippet(expr.span));
+            }
+
+            self.changes.push_str_span(field.span, &result);
+
+            if !last_field || config!(struct_trailing_comma) {
```
Comment (nrc) [inline src/items.rs]: we should have another config option for `enum_trailing_comma`

File: src/items.rs [hunk 7]
```diff
@@ -408,6 +408,122 @@ impl<'a> FmtVisitor<'a> {
...
+                self.changes.push_str_span(field.span, ",");
```
Comment (nrc) [inline src/items.rs]: Do we also need to take account of the comma in the types budgets?

File: tests/idem/enum.rs [hunk 8]
```diff
@@ -0,0 +1,33 @@
+// Enums test
+
+#[atrr]
+pub enum Test {
+    A,
+    B(u32, A /* comment */),
```
Comment (nrc) [inline tests/idem/enum.rs]: Could you add a test where the types in the variant get put on multiple lines?

File: src/items.rs [hunk 9]
```diff
@@ -408,6 +408,122 @@ impl<'a> FmtVisitor<'a> {
...
     }
 
+    pub fn visit_enum(&mut self,
+                      ident: ast::Ident,
+                      vis: ast::Visibility,
+                      enum_def: &ast::EnumDef,
+                      generics: &ast::Generics,
+                      span: Span)
+    {
+        // Empty enums are a bit weird (and not like empty structs), let's not touch them for now
```
Comment (nrc) [inline src/items.rs]: Why are they weird? I'd think you just format as:
```
enum Foo<...> {}
```

or

```
enum Foo<...> {
}
```

depending on an option. Is there something I'm missing?

File: src/items.rs [hunk 10]
```diff
@@ -408,6 +408,122 @@ impl<'a> FmtVisitor<'a> {
...
+        if enum_def.variants.len() == 0 {
+            let result = self.snippet(span);
+            self.changes.push_str_span(span, &result);
```
Comment (nrc) [inline src/items.rs]: You don't need to do this, you can just return and the span will get automatically added the next time we try to format something.

File: src/items.rs [hunk 11]
```diff
@@ -408,6 +408,122 @@ impl<'a> FmtVisitor<'a> {
...
+        }
+
+        let header_str = self.format_header("enum", ident, vis);
+        self.changes.push_str_span(span, &header_str);
+
+        let generics_buf = self.format_generics(generics, enum_def.variants[0].span.lo);
+        self.changes.push_str_span(span, &generics_buf);
+
+        let enum_snippet = self.snippet(span);
+        // FIXME this will give incorrect results if there is a { in a commet.
```
Comment (nrc) [inline src/items.rs]: typo: commet

File: src/items.rs [hunk 12]
```diff
@@ -408,6 +408,122 @@ impl<'a> FmtVisitor<'a> {
...
+        if enum_def.variants.len() == 0 {
+            let result = self.snippet(span);
+            self.changes.push_str_span(span, &result);
+            return;
+        }
+
+        let header_str = self.format_header("enum", ident, vis);
+        self.changes.push_str_span(span, &header_str);
+
+        let generics_buf = self.format_generics(generics, enum_def.variants[0].span.lo);
```
Comment (nrc) [inline src/items.rs]: nit: generics_str for consistency with header_str

File: src/items.rs (hunk 9 above)
Comment (marcusklaas) [inline src/items.rs]: You are correct. I was afraid of dropping comments, but they can easily be preserved.

File: tests/idem/enum.rs (hunk 8 above)
Comment (marcusklaas) [inline tests/idem/enum.rs]: There is one already: `LongVariants`, or is that not what you mean?

File: src/items.rs (hunk 7 above)
Comment (marcusklaas) [inline src/items.rs]: Yes! Fixed.

File: src/items.rs (hunk 5 above)
Comment (marcusklaas) [inline src/items.rs]: How would we format such a thing would it go over budget? I think that case is extremely uncommon, since `disr_expr` is only set for C-style enums, which are generally short. I will add an assert.

**Comment by marcusklaas** [comment]
Thanks for all the feedback!

File: tests/idem/enum.rs (hunk 8 above)
Comment (nrc) [inline tests/idem/enum.rs]: Oh yeah, perfect, didn't see that before
